source/uwp/Renderer/lib/XamlBuilder.cpp

```diff
@@ -829,9 +823,43 @@ AdaptiveNamespaceStart
                 XamlHelpers::SetContent(localButton.Get(), title.Get());
                 return;
             }
-            XamlHelpers::AppendXamlElementToPanel(buttonIcon.Get(), buttonContentsPanel.Get()); // Add image to stack panel
+
+            // Create title text block
+            ComPtr<ITextBlock> buttonText = XamlHelpers::CreateXamlClass<ITextBlock>(HStringReference(RuntimeClass_Windows_UI_Xaml_Controls_TextBlock));
+            THROW_IF_FAILED(buttonText->put_Text(title.Get()));
+            THROW_IF_FAILED(buttonText->put_TextAlignment(TextAlignment::TextAlignment_Center));
+
+            // Handle different arrangements inside button
             ComPtr<IFrameworkElement> buttonIconAsFrameworkElement;
             THROW_IF_FAILED(buttonIcon.As(&buttonIconAsFrameworkElement));
+            if (iconPlacement == ABI::AdaptiveNamespace::IconPlacement::AboveTitle && allActionsHaveIcons)
+            {
+                THROW_IF_FAILED(buttonContentsStackPanel->put_Orientation(Orientation::Orientation_Vertical));
+
+                // Set height to iconSize (aspect ratio is automatically maintained)
+                THROW_IF_FAILED(buttonIconAsFrameworkElement->put_Height(iconSize));
+            }
+            else
+            {
+                THROW_IF_FAILED(buttonContentsStackPanel->put_Orientation(Orientation::Orientation_Horizontal));
+
+                // Add event to the image to resize itself when the textblock is rendered
+                ComPtr<IImage> buttonIconAsImage;
+                THROW_IF_FAILED(buttonIcon.As(&buttonIconAsImage));
+
+                EventRegistrationToken eventToken;
+                THROW_IF_FAILED(buttonIconAsImage->add_ImageOpened(Callback<IRoutedEventHandler>(
+                    [buttonIconAsFrameworkElement, buttonText](IInspectable* /*sender*/, IRoutedEventArgs* /*args*/) -> HRESULT
+                {
+                    return SetImageSizeAsTextBlockSize(buttonIconAsFrameworkElement.Get(), buttonText.Get());
+                }).Get(), &eventToken));
+            }
+
+            ComPtr<IPanel> buttonContentsPanel;
+            THROW_IF_FAILED(buttonContentsStackPanel.As(&buttonContentsPanel));
+
+            // Add image to stack panel
+            XamlHelpers::AppendXamlElementToPanel(buttonIcon.Get(), buttonContentsPanel.Get());
 
             // Just add spacing when the icon must be located at the left of the title
             if (iconPlacement == ABI::AdaptiveNamespace::IconPlacement::LeftOfTitle)
```

Comment:
>iconPlacement == ABI::AdaptiveNamespace::IconPlacement::LeftOfTitle [](start = 16, length = 67)

Shouldn't the spacing be added when not al actions have icons also? Since that will result in LeftOfTitle behavior also? Shouldn't this block be in the if clause above?

Reply:
you're right. I didn't notice it since it was there before. Fixed in d8f2554